queue/queueservice.js

```diff
@@ -5,63 +5,280 @@ var debug       = require('debug')('queue:queue');
+    // Ensure the claim queue exists
+    await this.ensureClaimQueue();
+
+    // Get messages
+    var messages = await new Promise((accept, reject) => {
+      this.service.getMessages(this.claimQueue, {
+        numOfMessages:        32,
+        peekOnly:             false,
+        visibilityTimeout:    10 * 60,
+      }, (err, messages) => { err ? reject(err) : accept(messages) });
+    });
+
+    // Convert to neatly consumable format
+    return messages.map((message) => {
+      var payload = JSON.parse(message.messagetext);
+      return {
+        taskId:     payload.taskId,
+        runId:      payload.runId,
+        takenUntil: new Date(payload.takenUntil),
+        remove:     () => {
+          return new Promise((accept, reject) => {
+            this.service.deleteMessage(
+              this.claimQueue,
+              message.messageid, message.popreceipt,
+              (err) => { err ? reject(err) : accept() }
+            );
+          });
+        }
+      };
+    });
+  }
+
+  /**
+   * Poll deadline resolution queue, returns promise for list of message objects
+   * on the form:
+   *
+   * ```js
+   * [
+   *   {
+   *     taskId:      '<taskId>',     // Task to check
+   *     deadline:    [Date object],  // Deadline of task when submitted
+   *     remove:      function() {},  // Delete message call when handled
+   *   },
+   *   ... // up-to to 32 objects in one list
+   * ]
+   * ```
+   *
+   * Note, messages must be handled within 10 minutes.
+   */
+  async pollDeadlineQueue() {
+    // Ensure the deadline queue exists
+    await this.ensureDeadlineQueue();
 
-/** Ensure existence of a queue */
-QueueService.prototype.ensureQueue = function(provisionerId, workerType) {
-  // Construct id, note that slash cannot be used in provisionerId, workerType
-  var id = provisionerId + '/' + workerType;
+    // Get messages
+    var messages = await new Promise((accept, reject) => {
+      this.service.getMessages(this.deadlineQueue, {
+        numOfMessages:        32,
+        peekOnly:             false,
+        visibilityTimeout:    10 * 60,
+      }, (err, messages) => { err ? reject(err) : accept(messages) });
+    });
+
+    // Convert to neatly consumable format
+    return messages.map((message) => {
+      var payload = JSON.parse(message.messagetext);
+      return {
+        taskId:     payload.taskId,
+        deadline:   new Date(payload.deadline),
+        remove:     () => {
+          return new Promise((accept, reject) => {
+            this.service.deleteMessage(
+              this.deadlineQueue,
+              message.messageid, message.popreceipt,
+              (err) => { err ? reject(err) : accept() }
+            );
+          });
+        }
+      };
+    });
+  }
 
-  // Find promise
-  var retval = this.queues[id];
+  /** Ensure existence of a queue */
+  ensurePendingQueue(provisionerId, workerType) {
+    // Construct id, note that slash cannot be used in provisionerId, workerType
```

Comment:
`+ ensurePendingQueue(provisionerId, workerType) {'

is this valid syntax?  I'm not familiar with it if it is...